src/main/java/org/yourteamname/agents/BaseAgent.java

```diff
@@ -17,121 +20,152 @@
-        private MessageTemplate mt;
-        private BaseAgent ba;
-
-        public void action(){
-            this.mt = MessageTemplate.and(MessageTemplate.MatchPerformative(ACLMessage.INFORM),
-                    MessageTemplate.MatchSender(baseAgent.clockAgent));
-            ACLMessage msg = myAgent.receive(this.mt);
-            if (msg != null) {
-                String messageContent = msg.getContent();
-                int counter = Integer.parseInt(messageContent);
-                int day = counter / 24;
-                int hour = counter % 24;
-                currentDay = day;
-                currentHour = hour;
-                allowAction = true;
-            }
-            else {
-                block();
-            }
-        }
-   }
+	private int currentHour;
+	private boolean allowAction = false;
+	protected AID clockAgent = new AID("TimeKeeper", AID.ISLOCALNAME);
+	protected AID messageQueueAgent = new AID("MessageQueue", AID.ISLOCALNAME);
+	protected BaseAgent baseAgent = this;
+
+	/*
+	 * Setup to add behaviour to talk with clockAgent Call `super.setup()` from
+	 * `setup()` function
+	 */
+	protected void setup() {
+		this.addBehaviour(new PermitAction());
+	}
+
+	/*
+	 * This function registers the agent to yellow pages Call this in `setup()`
+	 * function
+	 */
+	protected void register(String type, String name) {
+		DFAgentDescription dfd = new DFAgentDescription();
+		dfd.setName(getAID());
+		ServiceDescription sd = new ServiceDescription();
+		sd.setType(type);
+		sd.setName(name);
+		dfd.addServices(sd);
+		try {
+			DFService.register(this, dfd);
+		} catch (FIPAException fe) {
+			fe.printStackTrace();
+		}
+	}
+
+	/*
+	 * This function removes the agent from yellow pages Call this in `doDelete()`
+	 * function
+	 */
+	protected void deRegister() {
+		try {
+			DFService.deregister(this);
+		} catch (FIPAException fe) {
+			fe.printStackTrace();
+		}
+	}
+
+	/*
+	 * This function sends finished message to clockAgent This function should be
+	 * called by every agent which implements BaseAgent after the agent is done with
+	 * the task it has to perform in a time step.
+	 */
+	protected void finished() {
+		this.allowAction = false;
+		ACLMessage finish = new ACLMessage(ACLMessage.INFORM);
+		finish.addReceiver(this.clockAgent);
+		finish.setContent("finished");
+		this.send(finish);
+	}
+
+	protected boolean getAllowAction() {
+		return allowAction;
+	}
+
+	protected int getCurrentDay() {
+		return currentDay;
+	}
+
+	protected int getCurrentHour() {
+		return currentHour;
+	}
+
+	/*
+	 * This function is used as a middle man which uses the message for different
+	 * visualisation methods Use `baseAgent.sendMessage(message)` instead of
+	 * `myAgent.send(message)` in every behaviour.
+	 */
+	protected void sendMessage(ACLMessage msg) {
+		this.send(msg);
+		this.visualiseHistoricalView(msg);
+		this.visualiseIndividualOrderStatus(msg);
+		this.visualiseMessageQueuesByAgent(msg, false);
```

Comment:
When an agent sends a message, it does not neccessarily mean that its message queue's amount should be reduced. For example, if an agent A has 5 messages in its queue from an agent B. Now it sends a message to another agent C. Here, the message queue for A is still 5 (it still has to service 5 messages from B), but as far as I understood the code, the message queue for A will be reduced to 4.

Reply:
I thought about that too but i wanted to ask the whole class when valuating it. It is not done completely yet. The problem is I think it will cause a bigger overhead for every agent if it has to provide the base agent access to an inner processing queue or list of the elements currently worked on. This list could then trigger the info on add and remove operations on itself. Since we did not discuss the structure of our agents in the class and agreed on one implementation for all, i could not really fix this problem. The lack of information forces me to just ignore this error for now and wait until we discuss all interfaces in class or I at least can see them all to check the implementation myself.
The javafx gui is not done yet but the message queue agent writes its status to console.